sys/malloc_monitor/malloc_monitor.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (C) 2024 TU Dresden
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @brief   monitor heap memory usage (calls to malloc/realloc/free)
+ * @author  Mikolai Gütschow <mikolai.guetschow@tu-dresden.de>
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include "architecture.h"
+#include "cpu.h"
+#include "malloc_monitor.h"
+#include "malloc_monitor_internal.h"
+
+#ifndef CONFIG_MODULE_SYS_MALLOC_MONITOR_SIZE
+#define CONFIG_MODULE_SYS_MALLOC_MONITOR_SIZE 100
+#endif
+
+#ifndef CONFIG_MODULE_SYS_MALLOC_MONITOR_VERBOSE
+#define CONFIG_MODULE_SYS_MALLOC_MONITOR_VERBOSE 0
+#endif
+
+static struct {
+    void *addr[CONFIG_MODULE_SYS_MALLOC_MONITOR_SIZE];
+    size_t size[CONFIG_MODULE_SYS_MALLOC_MONITOR_SIZE];
+    size_t current;
+    size_t high_watermark;
+} malloc_monitor = {
+    .addr = {NULL},
+    .current = 0,
+    .high_watermark = 0,
+};
+
+void malloc_monitor_add(void *ptr, size_t size, uinttxtptr_t pc, char *func_prefix)
+{
+    if (ptr == NULL) return;
```

Comment:
```suggestion
    if (ptr == NULL) {
        return;
    }
```

The coding convention wants us to always add {} blocks, no matter what. Given the "goto fail" bug, I personally see value in this.